bundles/org.openhab.binding.monopriceaudio/src/main/java/org/openhab/binding/monopriceaudio/internal/handler/MonopriceAudioHandler.java

```diff
@@ -0,0 +1,730 @@
+                            }
+                        }
+                        break;
+                    case CHANNEL_TYPE_BASS:
+                        if (command instanceof DecimalType) {
+                            int value = ((DecimalType) command).intValue();
+                            if (value >= MIN_TONE && value <= MAX_TONE) {
+                                logger.debug("Got bass command {} zone {}", value, zone);
+                                connector.sendCommand(zone, MonopriceAudioCommand.BASS, value + TONE_OFFSET);
+                                ZONE_DATA_MAP.get(zone.getZoneId()).setBass(String.format("%02d", value + TONE_OFFSET));
+                            }
+                        }
+                        break;
+                    case CHANNEL_TYPE_BALANCE:
+                        if (command instanceof DecimalType) {
+                            int value = ((DecimalType) command).intValue();
+                            if (value >= MIN_BALANCE && value <= MAX_BALANCE) {
+                                logger.debug("Got balance command {} zone {}", value, zone);
+                                connector.sendCommand(zone, MonopriceAudioCommand.BALANCE, value + BALANCE_OFFSET);
+                                ZONE_DATA_MAP.get(zone.getZoneId())
+                                        .setBalance(String.format("%02d", value + BALANCE_OFFSET));
+                            }
+                        }
+                        break;
+                    case CHANNEL_TYPE_DND:
+                        if (command instanceof OnOffType && command == OnOffType.ON) {
+                            connector.sendCommand(zone, MonopriceAudioCommand.DND_ON);
+                            ZONE_DATA_MAP.get(zone.getZoneId()).setDnd(ON);
+                        } else if (command instanceof OnOffType && command == OnOffType.OFF) {
+                            connector.sendCommand(zone, MonopriceAudioCommand.DND_OFF);
+                            ZONE_DATA_MAP.get(zone.getZoneId()).setDnd(OFF);
+                        }
+                        break;
+                    case CHANNEL_TYPE_ALLON:
+                        zoneStream.limit(numZones).forEach((zoneId) -> {
+                            if (!ignoreZones.contains(zoneId)) {
+                                try {
+                                    connector.sendCommand(MonopriceAudioZone.valueOf(ZONE + zoneId),
+                                            MonopriceAudioCommand.POWER_ON);
+                                    Thread.sleep(SLEEP_BETWEEN_CMD);
+                                    // reset the volume of each zone to allVolume
+                                    connector.sendCommand(MonopriceAudioZone.valueOf(ZONE + zoneId),
+                                            MonopriceAudioCommand.VOLUME, allVolume);
+                                    Thread.sleep(SLEEP_BETWEEN_CMD);
```

Comment:
Are you aware that multiple `handleCommand()`s can be invoked at the same time from different threads? This rate limit might not work, then.

Reply:
Yes but hopefully the All* commands will be used sparingly by the user. The idea is to give some advanced functionality to the user so they can affect all zones with one command and avoid having to write rules that loop over all zones. For example, the 'All Off' command can be triggered by a rule that runs when arming the security system to away mode to turn off the music when leaving. And the 'All Volume' & 'All Source' provide party mode functionality that is native to more advanced WHA systems. So ideally two different users would never be adjusting the All Volume at the same time.

Reply:
Ok, but then you can remove the second `Thread.sleep()` in every `case`, because they have no effect.

Reply:
I removed all the sleeps and the device controller seems to be coping with this ok.